torch/csrc/inductor/aoti_eager/kernel_holder.cpp

```diff
@@ -0,0 +1,283 @@
+  if (ivalue.isTensor()) {
+    ivalue_type = IValueType::Tensor;
+  } else if (ivalue.isTensorList()) {
+    ivalue_type = IValueType::TensorList;
+  } else if (ivalue.isOptionalTensorList()) {
+    ivalue_type = IValueType::OptionalTensorList;
+  } else if (ivalue.isScalar()) {
+    ivalue_type = IValueType::Scalar;
+  }
+
+  auto it = handlers_.find(ivalue_type);
+  if (it != handlers_.end()) {
+    return it->second(ivalue, device, inputs);
+  }
+
+  // Handle unsupported types or add a default handler
+  return false;
+}
+
+bool unpackTensors(
+    const torch::jit::Stack& stack,
+    const c10::Device& device,
+    std::vector<at::Tensor>& inputs) {
+  for (const auto& ivalue : stack) {
+    if (!HandleIValue(ivalue, device, inputs)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+} // namespace
+
+AOTIPythonKernelHolder::AOTIPythonKernelHolder(
+    py::object func,
+    c10::DispatchKey dispatch_key,
+    c10::string_view ns,
+    c10::string_view op_name,
+    c10::string_view op_overload_name,
+    bool is_symbolic)
+    : python_kernel_holder_(func, dispatch_key),
+      dispatch_key_(dispatch_key),
+      ns_(std::string(ns)),
+      op_name_(std::string(op_name)),
+      op_overload_name_(std::string(op_overload_name)),
+      is_symbolic_(is_symbolic),
+      has_fall_back_(func.ptr() != Py_None),
+      device_opt_(c10::Device(c10::dispatchKeyToDeviceType(dispatch_key_), 0)),
+      pyinterpreter_(getPyInterpreter()) {
+  (void)is_symbolic_; // Suppress unused variable warning
+}
+
+void AOTIPythonKernelHolder::operator()(
+    const c10::OperatorHandle& op,
+    c10::DispatchKeySet keyset,
+    torch::jit::Stack* stack) {
+  if (cache_lookup(op, keyset, stack)) {
+    cache_hit(op, keyset, stack);
+  } else {
+    cache_miss(op, keyset, stack);
+  }
+}
+
+bool AOTIPythonKernelHolder::cache_lookup(
+    const c10::OperatorHandle& op,
+    c10::DispatchKeySet keyset,
+    torch::jit::Stack* stack) {
+  // Always return false not and will add cache lookup later.
+  return false;
+}
+
+void AOTIPythonKernelHolder::cache_hit(
+    const c10::OperatorHandle& op,
+    c10::DispatchKeySet keyset,
+    torch::jit::Stack* stack) {
+  TORCH_CHECK(has_fall_back_, "AOTI kernel failed to run and no fall back");
+  python_kernel_holder_(op, keyset, stack);
+}
+
+void AOTIPythonKernelHolder::cache_miss(
+    const c10::OperatorHandle& op,
+    c10::DispatchKeySet keyset,
+    torch::jit::Stack* stack) {
+  auto device_type = c10::dispatchKeyToDeviceType(dispatch_key_);
+  TORCH_CHECK(
+      (device_type == c10::DeviceType::CPU) ||
+          (device_type == c10::DeviceType::CUDA),
+      "Unsupported device type");
+
+  auto kernel_lib_path = produce_aoti_kernel_lib(op, keyset, stack);
+  if (kernel_lib_path.empty()) {
+    TORCH_CHECK(
+        has_fall_back_,
+        "Failed to produce kernel libarary by using AOTI and no fall back");
+    python_kernel_holder_(op, keyset, stack);
+    return;
```

Comment:
Are we hitting this in practice?  Should this just be an error?

Reply:
The registration API(`_impl_with_aoti_compile`) provides a fallback function input in case AOTI could not produce kernel. For example, users could provide a numpy implementation for a particular aten operation if AOTI failed to produce kernel. Do you think we do not need to provide such a fallback function? @jansel , @desertfire